x/wasm/IBC.md

```diff
@@ -0,0 +1,425 @@
+    // otherwise, this is just interesting metadata
+
+	// block height after which the packet times out 
+	TimeoutHeight uint64
+	// block timestamp (in nanoseconds) after which the packet times out
+	TimeoutTimestamp uint64
+}
+```
+
+Channel Lifecycle:
+
+```go
+package lifecycle
+
+// if this returns error, we reject the channel opening
+// otherwise response has the versions we accept
+//
+// It is provided the full ChannelInfo being proposed to do any needed checks
+func OnChanNegotiate(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, request ChannelInfo) 
+    (version Version, err error) {}
+
+// This is called with the full ChannelInfo once the other side has agreed.
+// An error here will still abort the handshake process. (so you can double check the order/version here)
+//
+// The main purpose is to allow the contract to set up any internal state knowing the channel was established,
+// and keep a registry with that ChannelID
+func OnChanOpened(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, request ChannelInfo) error {}
+
+// This is called when the channel is closed for any reason
+// TODO: any meaning to return an error here? we cannot abort closing a channel
+func OnChanClosed(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, request ChannelClosedInfo) error {}
+
+type ChannelInfo struct {
+    // key info to enforce (error if not what is expected)
+    Order channeltypes.Order
+    // the version info the counterparty contract is proposing / agreed to
+    CounterpartyVersion Version
+    // local id for the Channel that is being initiated
+    ChannelID string
+    // these two are taken from channeltypes.Counterparty
+    RemotePortID string
+    RemoteChannelID string
+}
+
+type ChannelClosedInfo struct {
+    // local id for the Channel that is being shut down
+    ChannelID string
+}
+```
+
+Queries:
+
+These are callbacks that the contract can make, calling into a QueryPlugin.
+The general type definition for a `QueryPlugin` is
+`func(ctx sdk.Context, request *wasmTypes.IBCQuery) ([]byte, error)`.
+All other info (like the contract address we are querying for) must be passed in
+from the contract (which knows it's own address).
+
+Here we just defined the request and response types (which will be serialized into `[]byte`)
+
+```go
+package queries
+
+type QueryPort struct {
+   ContractAddress string
+}
+
+type QueryPortResponse struct {
+   PortID string
+}
+
+type QueryChannels struct {
+    // exactly one of these must be set. ContractAddress is a shortcut to save the Contract->PortID mapping
+    PortID string
+	ContractAddress string
+}
+
+type QueryChannelsResponse struct {
+   ChannelIDs []ChannelMetadata
+}
+
+type ChannelMetadata struct {
+    // Local portID, channelID is our unique identifier
+    PortID string
+    ChannelID string
+    RemotePortID string
+    Order channeltypes.Order
+    CounterpartyVerson Version
```

Comment:
The `channel.version` is not fully clear to me. Either it is the counterparty version, the local one or both as they are the same. Go doc states: " opaque channel version, which is agreed upon during the handshake"

Reply:
> During the handshake process, two ends of a channel come to agreement on a version bytestring associated with that channel. 

https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics#versioning

Reply:
You are right, I was confused with `ChannelInfo` above (which is used *during* the negotiation process).

Once established, there is just a Version.